cavern/src/main/java/org/opencadc/cavern/files/PutAction.java

```diff
@@ -227,8 +224,27 @@ public void doAction() throws Exception {
                 throw e;
             }
         } finally {
-            log.debug("put: done " + nodeURI.getURI().toASCIIString());
+            if (successful) {
+                log.debug("put: done " + nodeURI.getURI().toASCIIString());
+            } else if (putStarted) {
+                cleanUpOnFailure(target, node, rootPath);
+            }
+        }
+    }
+    
+    private void cleanUpOnFailure(Path target, Node node, Path rootPath) throws IOException {
+        log.debug("clean up on put failure " + target);
+        Files.delete(target);
+        // restore empty DataNode: remove props that are no longer applicable
+        Iterator<NodeProperty> i = node.getProperties().iterator();
+        while (i.hasNext()) {
+            NodeProperty np = i.next();
+            if (VOS.PROPERTY_URI_CONTENTMD5.equals(np.getPropertyURI())) {
+                i.remove();
+            }
         }
+        NodeUtil.create(rootPath, node);
```

Comment:
I know this was here before, but why do we create the node again?

Reply:
Yeah, I am just refactoring the existing code. Since we won't be running extensive tests on the updated PutAction, I tried not to make significant changes.

Reply:
If PutAction is always invoked as a result of a adding a node, then we can remove the NodeUtil.create(...) statement. If PutAction can be invoked as a result of an node update, then we need to keep the statement.

Reply:
Keeping this statement to support the (to be implemented) 'feature' to allow uploads without having to create the node first.